core/arch/mangle_shared.c

```diff
@@ -801,27 +847,59 @@ mangle_rseq(dcontext_t *dcontext, instrlist_t *ilist, instr_t *instr, instr_t *n
         }
 #    ifdef X86
         /* We just ran the instrumented version of the rseq code, with the stores
-         * removed.  Now we need to invoke it again natively for real.  We have to
-         * invoke the abort handler, not the rseq start, as it may perform some setup.
+         * removed.  Now we need to invoke it again natively for real.  We would prefer
+         * to invoke the abort handler, as it may perform some setup, but in too many
+         * cases it is truly an "abort" handler that just exits rather than a "restart
+         * handler".  Furthermore, to support executing a copy of the code natively in
+         * order to provide guarantees on regaining control and not rely on call-return
+         * semantics, it is simpler to execute only the limited-scope rseq region.
+         * Thus, we target the start point.
+         *
+         * In case the abort handler does perform setup, we checkpoint and restore GPR
+         * register values.  Memory should remain as it was, due to nop-ing of stores.
          *
-         * XXX i#2350: We assume the code can handle "aborting" once every time and
-         * still make forward progress.  We document this in our Limitations section.
-         * We also ignore the flags for when to restart.  It's possible the app
+         * XXX i#2350: We ignore the flags for when to restart.  It's possible the app
```

Comment:
Could you re-phrase? Which flags? Also: It's possible the app here <what>?

Reply:
It says "the flags for when to restart".  Each rseq region has flags indicating when its handler should be called: on a preempt, migrate, and/or signal.  The only thing I can think of is to add "app's" in front of "flags"?

Reply:
Or rseq region's [..]
Ok this wasn't clear.